### DISCUSSION TITLE
supporting org `file:` links in `company-files`

### DISCUSSION THREAD

**Discussion by martianhi** [company-mode/company-mode#1319] [Ideas]
org has a quick file link creation ability: if you type `file:` followed by some path you can make a link without using the `[[link][description]]` syntax, but it doesn't offer completion.

i thought it would be nice if `company-files` could work with this, so you can quickly insert file links with completion, without having to do it the long way (ie `C-c C-l`, `file:` `RET` choosing file, then typing a description).

i couldn't see that `company` already does this, so i hacked up some code, but it probably needs improving by someone who knows company itself. i noticed that there's no need to add the `[[` to the regexp, `file:` alone works, whether you type `[[file:` or just `file:`.

i suspect that modifying `company-files--regexps` with a mode-specific regex is a bad idea, and i'm unsure of the alternative, so we just test for the major mode first:

```emacs-lisp

(defun company-files--grab-existing-name ()
  ;; Grab the file name.
  ;; When surrounded with quotes, it can include spaces.
  (let (file dir)
    (when (eq major-mode 'org-mode)
      (company-files--mod-regexps-for-org-links))
    (and (cl-dolist (regexp company-files--regexps)
           (when (setq file (company-grab-line regexp 1))
             (cl-return file)))
         (if (string-prefix-p "file:" file) ; org file links
             (setq file (substring-no-properties file 5))
           file)
         (company-files--connected-p file)
         (setq dir (file-name-directory file))
         (not (string-match "//" dir))
         (file-exists-p dir)
         file)))

(defun company-files--mod-regexps-for-org-links ()
  "Modify `company-files--regexps' to work on org mode's 'file:' links."
  (setq company-files--regexps
        (let* ((root (if (eq system-type 'windows-nt)
                         "[a-zA-Z]:/"
                       "/"))
               (begin (concat "\\(?:\\.\\{1,2\\}/\\|~/\\|" root "\\)"))
               (org-file-prefix "\\(file:\\)?")) ;org file links
          (list (concat "\"\\(" org-file-prefix begin "[^\"\n]*\\)")
                (concat "\'\\(" org-file-prefix begin "[^\'\n]*\\)")
                (concat "\\(?:[ \t=\[]\\|^\\)\\(" org-file-prefix begin "[^ \t\n]*\\)")))))
```

or perhaps i am missing something and this is already possible some other way?

**Comment by yugaego** [comment]
From my experience, the obstacle here was that completion was [not enabled from inside a symbol](https://github.com/company-mode/company-mode/issues/340). Do I get you right you're referring to this issue too?

One more thing that could probably bother you was discussed in the [file completion without leading slashes](https://github.com/company-mode/company-mode/issues/668) thread.
